src/compiler/transformers/es2015.ts

```diff
@@ -3262,6 +3267,147 @@ namespace ts {
+
+            // Under certain conditions, the 'ts' transformer may may introduce a class alias, which
+            // we see as an assignment, for example:
+            //
+            //  (function () {
+            //      var C = C_1 = (function () {
+            //          function C() {
+            //          }
+            //          C.x = function () { return C_1; }
+            //          return C;
+            //      }());
+            //      C = C_1 = __decorate([dec], C);
+            //      return C;
+            //      var C_1;
+            //  }())
+            //
+            const aliasAssignment = isAssignmentExpression(initializer) ? initializer : undefined;
+
+            // The underlying call (3) is another IIFE that may contain a '_super' argument.
+            const call = cast(aliasAssignment ? skipOuterExpressions(aliasAssignment.right) : initializer, isCallExpression);
+            const func = cast(skipOuterExpressions(call.expression), isFunctionExpression);
+
+            // When we extract the statements of the inner IIFE, we exclude the 'return' statement (4)
+            // as we already have one that has been introduced by the 'ts' transformer.
+            const funcStatements = func.body.statements.slice(0, -1);
+
+            const statements: Statement[] = [];
+            if (aliasAssignment) {
+                // If we have a class alias assignment, we need to move it to the down-level constructor
+                // function we generated for the class.
+                const hasExtendsCall = isExpressionStatement(funcStatements[0]);
+                if (hasExtendsCall) {
+                    statements.push(funcStatements[0]);
+                }
+
+                // We reuse the comment and source-map positions from the original variable statement
+                // and class alias, while converting the function declaration for the class constructor
+                // into an expression.
+                statements.push(
+                    updateVariableStatement(
+                        varStatement,
+                        /*modifiers*/ undefined,
+                        updateVariableDeclarationList(varStatement.declarationList, [
+                            updateVariableDeclaration(variable,
+                                variable.name,
+                                /*type*/ undefined,
+                                updateBinary(aliasAssignment,
+                                    aliasAssignment.left,
+                                    convertFunctionDeclarationToExpression(
+                                        cast(funcStatements[hasExtendsCall ? 1 : 0], isFunctionDeclaration)
+                                    )
+                                )
+                            )
+                        ])
+                    )
+                );
+
+                addRange(statements, funcStatements.slice(hasExtendsCall ? 2 : 1));

```

Comment:
`addRange` will try filtering out `undefined` elements from `funcStatements` -- if this isn't necessary, I think we should just have a `pushMany` helper. It could also support pushing a slice so you don't have to `addRange(foo, bar.slice(...))`.